File: arrays/first-missing-positive/first-missing-positive.js
/**
 * @param {number[]} nums
 * @return {number}
 */
function firstMissingPositive(nums) {
  const n = nums.length;

  for (let i = 0; i < n; i++) {
    if (nums[i] <= 0 || nums[i] > n) {
      nums[i] = n + 1;
    }
  }

  for (let i = 0; i < n; i++) {
    let num = Math.abs(nums[i]);

    if (num > n) {
      continue;
    }

    if (nums[num - 1] > 0) {
      nums[num - 1] = -1 * nums[num - 1];
    }
  }

  for (let i = 0; i < n; i++) {
    if (nums[i] > 0) {
      return i + 1;
    }
  }
  return n + 1;
}

File: arrays/first-missing-positive/first-missing-positive.go
package firstmissingpositive

func firstMissingPositive(nums []int) int {
	n := len(nums)

	for i := 0; i < n; i++ {
		if nums[i] <= 0 || nums[i] > n {
			nums[i] = n + 1
		}
	}

	for i := 0; i < n; i++ {
		num := nums[i]

		if num < 0 {
			num = -num
		}

		if num > n {
			continue
		}

		if nums[num-1] > 0 {
			nums[num-1] = -1 * nums[num-1]
		}
	}

	for i := 0; i < n; i++ {
		if nums[i] > 0 {
			return i + 1
		}
	}
	return n + 1
}
